### PR TITLE
5591 cached vizjson behind wrong url

### LINKED ISSUE #5591: Cached vizjson behind wrong URL
Since #4360 vizjson urls must have the username, otherwise [visualizations controller returns 404](https://github.com/CartoDB/cartodb/pull/4394/files#diff-b6f9b42614afc35530e3525a33a64b4eR147). Nevertheless, there's at least one wrong url working: `https://team.cartodb.com/api/v2/viz/b903bb8e-255b-11e5-abcd-0e4fddd5de28/viz.json`. It's cached (appending `?wadus` displays the right error response), but it shouldn't have been generated. 

Just discovered it's related to sessions.

Example:
- Right url: https://team.cartodb.com/u/juanignaciosl/api/v2/viz/d0df01de-4917-11e5-92da-0e9d821ea90d/viz.json
- Wrong url: https://team.cartodb.com/api/v2/viz/d0df01de-4917-11e5-92da-0e9d821ea90d/viz.json

The wrong url won't work in a browser window without authenticated user, but works if you open it with an authenticated user.

cc @CartoDB/backend @javisantana

### PR DESCRIPTION
**Pull request by juanignaciosl**
This fixes #5591

### CODE REVIEW

**Comment by juanignaciosl** [comment]
Retest this, please.

**Comment by juanignaciosl** [comment]
Retest this, please.

**Comment by juanignaciosl** [comment]
Retest this, please.

**Comment by jesusvazquez** [comment]
retest this please

**Comment by juanignaciosl** [comment]
@Kartones CR this, please. Take a look at test battery as well, maybe you miss some case that might be relevant.

File: app/controllers/carto/api/visualizations_controller.rb [hunk 1]
```diff
@@ -7,6 +7,15 @@
 
 module Carto
   module Api
+    class VisualizationLoadError < StandardError
```
Comment (Kartones) [inline app/controllers/carto/api/visualizations_controller.rb]: I'd add exceptions below the main class, not above, but is personal taste

File: app/controllers/carto/api/visualizations_controller.rb [hunk 2]
```diff
@@ -145,11 +156,28 @@ def load_visualization
...
-        subdomain = CartoDB.extract_subdomain(request)
-        # INFO: subdomain checking is not part of previous permission checking (and should not be), so we add an additional check here: subdomain must match visualization username.
-        return render(text: 'Visualization of that user does not exist', status: 404) if subdomain && !subdomain.empty? && subdomain != @visualization.user.username && !@visualization.has_read_permission?(current_viewer)
+        raise Carto::Api::VisualizationLoadError.new('Visualization does not exist', 404) if @visualization.nil?
+        raise Carto::Api::VisualizationLoadError.new('Visualization not viewable', 403) if !@visualization.is_viewable_by_user?(current_viewer)
+        raise Carto::Api::VisualizationLoadError.new('Visualization of that user does not exist', 404) unless request_username_matches_visualization_owner
+        # Check for previous line before #5591 (in case fix doesn't cover every url)
+        # subdomain = CartoDB.extract_subdomain(request) # needed for check
+        # if subdomain && !subdomain.empty? && subdomain != @visualization.user.username
```
Comment (Kartones) [inline app/controllers/carto/api/visualizations_controller.rb]: Apart from removing the commented code, I dislike not doing an explicit `@visualization` at the end here, as there are multiple raises and is not crystal clear that "you're going to return a visualization if all went well" (no matter what style guides say)...

File: spec/factories/organizations_contexts.rb [hunk 3]
```diff
@@ -80,15 +80,35 @@ def share_table_with_organization(table, owner, organization)
...
+                       entity: { id: organization.id },
+                       access: CartoDB::Permission::ACCESS_READONLY
+                     }]
+             }
+    url = api_v1_permissions_update_url(user_domain: owner.username, api_key: owner.api_key, id: perm_id)
+    put url, params.to_json, headers
     last_response.status.should == 200
   end
 
+  def share_visualization(visualization, user)
```
Comment (Kartones) [inline spec/factories/organizations_contexts.rb]: `share_visualization_as_readonly` maybe to be explicit about type of permission granted? other than that, nice

File: app/controllers/carto/api/visualizations_controller.rb [hunk 4]
```diff
@@ -145,11 +156,28 @@ def load_visualization
...
+      # Remember that the url of a visualization shared with a user contains that user's username instead of owner's
+      def request_username_matches_visualization_owner
+        # Support both for username at `/u/username` and subdomain, prioritizing first
+        username = [CartoDB.username_from_request(request), CartoDB.subdomain_from_request(request)].compact.first
+        # URL must always contain username, either at subdomain or at path.
+        # Domainless url documentation: http://cartodb.readthedocs.org/en/latest/configuration.html#domainless-urls
+        return false unless username.present?
+
+        # Either user is owner or is current and has permission
+        # R permission check is based on current_viewer because current_user assumes you're viewing your subdomain
```
Comment (Kartones) [inline app/controllers/carto/api/visualizations_controller.rb]: ok, this one is hard to shallow but after some tests I think everything is right.

current_user is only not null if you're viewing your own subdomain but as you cannot enter shared writeable stuff except if using "your domain share url" (the one with `tables/USER.TABLENAME`), fine

**Comment by Kartones** [comment]
I would add a spec or two where `current_user` and `current_viewer` are not the same, because you have only scenarios where both are the same and that's where in the past we've had issues.

If they differ, either `current_user` is the owner of the vis, or `current_viewer` should have permissions, so just with org-ones but I'd add scenarios for that.

apart from that... +1

File: app/controllers/carto/api/visualizations_controller.rb (hunk 1 above)
Comment (juanignaciosl) [inline app/controllers/carto/api/visualizations_controller.rb]: I would as well, but Ruby interpreter didn't catch it unless I moved it up :( And I'd like to keep it only at that file, since it's meant to be used only at that controller.

As a side note, we should slowly move from rendering in filters to raising errors and using `rescue_from`, cleaner and Rails 4 friendly.

File: app/controllers/carto/api/visualizations_controller.rb (hunk 2 above)
Comment (juanignaciosl) [inline app/controllers/carto/api/visualizations_controller.rb]: Uhm. Take into account this is a `before_filter`, it's not expected to _return_ anything but to _load_ an instance variable (I think I named this `load_` instead of `get_` because of this). Anyway, if you feel it's clearer with a return value in spite of being a filter, I'll add it.

File: spec/factories/organizations_contexts.rb (hunk 3 above)
Comment (juanignaciosl) [inline spec/factories/organizations_contexts.rb]: Ok, `def share_visualization(visualization, user, access = CartoDB::Permission::ACCESS_READONLY)`, both explicit and flexible :-)

File: app/controllers/carto/api/visualizations_controller.rb (hunk 2 above)
Comment (juanignaciosl) [inline app/controllers/carto/api/visualizations_controller.rb]: PD: commented code deleted as well.

File: app/controllers/carto/api/visualizations_controller.rb (hunk 1 above)
Comment (Kartones) [inline app/controllers/carto/api/visualizations_controller.rb]: like it

File: app/controllers/carto/api/visualizations_controller.rb (hunk 2 above)
Comment (Kartones) [inline app/controllers/carto/api/visualizations_controller.rb]: ahh right, the `render` that was before as last line mislead me, it is setting a `@visualization` variable. You're absolutely right!

**Comment by jesusvazquez** [comment]
retest this please

**Comment by jesusvazquez** [comment]
retest this please

**Comment by jesusvazquez** [comment]
retest this please

**Comment by juanignaciosl** [comment]
@Kartones I've added some tests that I think cover the scenarios you suggested. I've also tested (currently in ded02) and it looks fine. Take a look at the tests and maybe a second CR if you want. If everything's ok I'll deploy tomorrow morning warning support for them to be aware.

**Comment by Kartones** [comment]
:+1:

**Comment by juanignaciosl** [comment]
Retest this, please.

**Comment by juanignaciosl** [comment]
Retest this, please.

**Comment by juanignaciosl** [comment]
Retest this, please.

**Comment by juanignaciosl** [comment]
@Kartones , @jesusvazquez , final CR, please. As you see the fixes for the tests have been quite a few, sorry for repeating the CR. It's currently deployed in ded02, it'd be great if you tested some corner case or strange usage you might have in mind :-)

**Comment by Kartones** [comment]
+1 especially since it gets green light. Good job!

**Comment by jesusvazquez** [comment]
+1 Everything seems to be OK.
